### PR TITLE
add cache in the graph and calendar

### PR DESCRIPTION
**Pull request by Antonio06188**

### CODE REVIEW

File: app/src/main/java/ch/epfl/sdp/healthplay/Frag_Home.java [hunk 1]
```diff
@@ -210,4 +212,21 @@ private void constructUserStats(FirebaseUser user){
         }
     }
 
+    private Map<String, Map<String, String>> convert(){
```
Comment (Zephyr75) [inline app/src/main/java/ch/epfl/sdp/healthplay/Frag_Home.java:215]: Some missing comments in the convert method

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutput;
+import java.io.ObjectOutputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DataCache {
```
Comment (Zephyr75) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:31]: Clever cache implementation, please add some comments

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    public void init(Context context){
+        this.context = context;
+        fa.addAuthStateListener(new FirebaseAuth.AuthStateListener() {
+            @Override
+            public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
+                FirebaseUser user = fa.getCurrentUser();
+                if(user==null)
+                    read();
+                else {
+                    db.mDatabase.child("users").child(user.getUid()).child("stats").addValueEventListener(new ValueEventListener() {
```
Comment (Zephyr75) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:55]: Avoid using magic numbers here

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                        }
+                    });
+                    store(user.getUid());
+                }
+            }
+        });
+    }
+
+    private String readFile() throws IOException{
+        ObjectInputStream in = new ObjectInputStream(new FileInputStream(new File(new File(context.getCacheDir(),"")+"cacheFile.txt")));
```
Comment (Zephyr75) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:73]: Avoid using magic strings for cache file name

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            } else {
+                Map<String, Map<String, Number>> map = (Map<String, Map<String, Number>>) task.getResult().getValue();
+                current_map = map;
+                ObjectMapper mapper = new ObjectMapper();
+                TypeReference<HashMap<String, HashMap<String, Number>>> typeRef = new TypeReference<HashMap<String, HashMap<String, Number>>>(){};
+                ObjectOutputStream out = null;
+                try {
+                    String json = mapper.writerFor(typeRef).writeValueAsString(map);
+                    Log.d("Ecriture", json);
+                    out = new ObjectOutputStream(new FileOutputStream(new File(context.getCacheDir(),"")+"cacheFile.txt"));
```
Comment (Zephyr75) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:109]: Also avoid using magic strings

**Review by Zephyr75 — CHANGES_REQUESTED** [review]
Good implementation overall, some missing comments and magic numbers but your cache should be easy to use!

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+                        }
+                    });
+                    store(user.getUid());
+                }
+            }
+        });
+    }
+
+    private String readFile() throws IOException{
```
Comment (Tetardfurieux) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:72]: Comment it if tho it's self explanatory, but explain how it works inside the function, like "This line open a Stream ect..."

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        StringBuilder stringBuilder = new StringBuilder();
+        while((character = in.read()) != -1){
+            stringBuilder.append((char)character);
+        }
+        Log.d("Debug lecture du fichier : ", stringBuilder.toString());
+        in.close();
+        return stringBuilder.toString();
+    }
+
+    private void read(){
```
Comment (Tetardfurieux) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:85]: Same as last function, describe the function with the /** **/ block, and also what happens inside

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            String jsonInput = readFile();
+            ObjectMapper mapper = new ObjectMapper();
+            TypeReference<HashMap<String, HashMap<String, Number>>> typeRef = new TypeReference<HashMap<String, HashMap<String, Number>>>(){};
+            Map<String, Map<String, Number>> map = mapper.readValue(jsonInput, typeRef);
+            current_map = map;
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+    private void store(String id){
```
Comment (Tetardfurieux) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:96]: Same, comment it please

File: app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java (hunk 2 above)
Comment (Tetardfurieux) [inline app/src/main/java/ch/epfl/sdp/healthplay/database/DataCache.java:31]: /** **/ block missing that explains briefly what you created this class for

**Review by Tetardfurieux — CHANGES_REQUESTED** [review]
Good code overall, great feature !

However, please add comments to your functions with a /** **/ block + // insinde the function, so that we can understand it quickly without reading in depth the function

**Review by Zephyr75 — APPROVED** [review]

**Review by Tetardfurieux — APPROVED** [review]
Nice fix ! Now the code is readable

**Pull request merged**